File: sim/core/reforging.go
package core

import (
	"math"
	"slices"

	"github.com/wowsims/wotlk/sim/core/proto"
	"github.com/wowsims/wotlk/sim/core/stats"
)

// MaxItemProtoStats is the worldserver's MAX_ITEM_PROTO_STATS. mod-reforging refuses an item whose
// StatsCount is 0 or at least this, so only 1 to 9 stats can be reforged.
const MaxItemProtoStats = 10

// reforgeStatTypeToStats turns an item_template ItemModType into the sim stats it lands on. It has
// to agree with the item database's conversion (AddItemMod in tools/database/azerothcore), because
// a reforge moves the template's value and the sim moves it between the stats that value became:
// generic hit, crit and haste ratings count for both melee and spell, and attack power also counts
// as ranged attack power. A type missing here is one the sim has no stat for, and a reforge naming
// it does nothing rather than inventing or losing a stat.
var reforgeStatTypeToStats = map[int32][]stats.Stat{
	0:  {stats.Mana},
	1:  {stats.Health},
	3:  {stats.Agility},
	4:  {stats.Strength},
	5:  {stats.Intellect},
	6:  {stats.Spirit},
	7:  {stats.Stamina},
	12: {stats.Defense},
	13: {stats.Dodge},
	14: {stats.Parry},
	15: {stats.Block},
	16: {stats.MeleeHit},
	17: {stats.MeleeHit},
	18: {stats.SpellHit},
	19: {stats.MeleeCrit},
	20: {stats.MeleeCrit},
	21: {stats.SpellCrit},
	28: {stats.MeleeHaste},
	29: {stats.MeleeHaste},
	30: {stats.SpellHaste},
	31: {stats.MeleeHit, stats.SpellHit},
	32: {stats.MeleeCrit, stats.SpellCrit},
	35: {stats.Resilience},
	36: {stats.MeleeHaste, stats.SpellHaste},
	37: {stats.Expertise},
	38: {stats.AttackPower, stats.RangedAttackPower},
	39: {stats.RangedAttackPower},
	42: {stats.SpellPower},
	43: {stats.MP5},
	44: {stats.ArmorPenetration},
	45: {stats.SpellPower},
	47: {stats.SpellPenetration},
	48: {stats.BlockValue},
}

// ReforgeStats returns the stat change a reforge makes on the item, or zero stats when the server
// wouldn't allow it. reforging is the raid's mod-reforging config, from Character.Server(); nil
// means the live server's.
//
// It follows ItemReforge::IsReforgeable and ::Reforge, which read item_template rather than what the
// sim made of it: the source has to be one of the item's own stats, and the target one it doesn't
// have. So a rating only an equip spell grants is no source, and neither is a melee-only or
// spell-only rating standing in for the combined one.
func ReforgeStats(item *Item, reforge *proto.ItemReforge, reforging *Reforging) stats.Stats {
	var delta stats.Stats
	if item == nil || reforge == nil {
		return delta
	}
	if reforging == nil {
		reforging = LiveReforging()
	}
	if !reforging.Enabled || !reforging.reforgeableItem(item) {
		return delta
	}
	if !reforging.ReforgeableStatPair(reforge.FromStatType, reforge.ToStatType) {
		return delta
	}

	fromValue := item.serverStatValue(reforge.FromStatType)
	// the server refuses a target stat the item already has
	if fromValue <= 0 || item.serverStatValue(reforge.ToStatType) > 0 {
		return delta
	}
	fromStats, toStats := reforgeStatTypeToStats[reforge.FromStatType], reforgeStatTypeToStats[reforge.ToStatType]
	if len(fromStats) == 0 || len(toStats) == 0 {
		return delta
	}

	amount := reforging.Amount(fromValue)
	if amount < 1 {
		return delta
	}
	for _, stat := range fromStats {
		delta[stat] -= amount
	}
	for _, stat := range toStats {
		delta[stat] += amount
	}
	return delta
}

// CanReforge says whether the server would let this reforge sit on the item. It's the same rule
// ReforgeStats applies, so callers that only need the yes/no can't drift from it.
func CanReforge(item *Item, reforge *proto.ItemReforge, reforging *Reforging) bool {
	return ReforgeStats(item, reforge, reforging) != stats.Stats{}
}

// Amount is what CalculateReforgePct moves off a template stat of that value: the configured
// percentage of it, floored, in float32 like the server.
func (r *Reforging) Amount(statValue int32) float64 {
	if statValue <= 0 {
		return 0
	}
	return math.Floor(float64(float32(statValue) * (float32(r.Percentage) / 100)))
}

// ReforgeableStatPair says whether the config's stat list allows this pair of ItemModTypes at all.
// It's the half of the rule that doesn't need the item; ReforgeStats applies the rest.
func (r *Reforging) ReforgeableStatPair(fromStatType, toStatType int32) bool {
	return fromStatType != toStatType &&
		slices.Contains(r.StatTypes, fromStatType) && slices.Contains(r.StatTypes, toStatType)
}

// reforgeableItem is ItemReforge::IsReforgeable's item half: a stat list to reforge with, and 1 to
// 9 template stats. Its quality check needs no counterpart here, since the item database only
// carries ServerStats for items whose item_template row holds their stats, and heirlooms, the only
// equippable quality above legendary, scale theirs.
func (r *Reforging) reforgeableItem(item *Item) bool {
	return len(r.StatTypes) > 0 &&
		len(item.ServerStats) > 0 && len(item.ServerStats) < MaxItemProtoStats
}

// serverStatValue is what LoadItemStatInfo plus FindItemStat give for a stat type: the first
// item_template row of that type with a value above zero, else 0.
func (item *Item) serverStatValue(statType int32) int32 {
	for _, stat := range item.ServerStats {
		if stat.Type == statType && stat.Value > 0 {
			return stat.Value
		}
	}
	return 0
}

File: ui/core/proto_utils/reforging.ts
import {
	LIVE_SERVER_DEFAULTS,
	MAX_REFORGEABLE_STAT_TYPES,
	REFORGE_DEFAULT_PERCENTAGE,
	REFORGE_MAX_PERCENTAGE,
	REFORGE_MIN_PERCENTAGE,
} from '../constants/server_defaults_auto_gen.js';
import { ItemReforge, ServerSettings } from '../proto/common.js';
import { UIItem as Item } from '../proto/ui.js';

// MAX_ITEM_PROTO_STATS: mod-reforging refuses an item whose StatsCount is 0 or at least this.
const MAX_ITEM_PROTO_STATS = 10;

// Reforging is mod-reforging's config as the sim resolves it; core.Reforging in Go.
export interface Reforging {
	enabled: boolean;
	// Percentage of the source stat a reforge moves, e.g. 40.
	percentage: number;
	// ItemModType ids a reforge can take from or give to.
	statTypes: Array<number>;
}

// reforgingFor resolves an encounter's settings over the live server's, the same way newReforging
// does in sim/core/server_settings.go.
export function reforgingFor(settings?: ServerSettings | null): Reforging {
	const live = LIVE_SERVER_DEFAULTS.reforge;
	const reforge = settings?.reforge;

	let percentage = reforge?.percentage ?? live?.percentage ?? REFORGE_DEFAULT_PERCENTAGE;
	if (percentage < REFORGE_MIN_PERCENTAGE || percentage > REFORGE_MAX_PERCENTAGE) {
		percentage = REFORGE_DEFAULT_PERCENTAGE;
	}

	let statTypes = reforge?.statTypes?.length ? reforge.statTypes : live?.statTypes ?? [];
	if (statTypes.length > MAX_REFORGEABLE_STAT_TYPES) {
		statTypes = [];
	}

	return { enabled: reforge?.enable ?? live?.enable ?? false, percentage, statTypes: [...statTypes] };
}

export const LIVE_REFORGING = reforgingFor();

// mod-reforging's StatTypeToString, for the ItemModTypes a config can list.
const reforgeStatTypeNames: Record<number, string> = {
	0: 'Mana',
	1: 'Health',
	3: 'Agility',
	4: 'Strength',
	5: 'Intellect',
	6: 'Spirit',
	7: 'Stamina',
	12: 'Defense',
	13: 'Dodge',
	14: 'Parry',
	15: 'Block',
	16: 'Melee Hit',
	17: 'Ranged Hit',
	18: 'Spell Hit',
	19: 'Melee Crit',
	20: 'Ranged Crit',
	21: 'Spell Crit',
	28: 'Melee Haste',
	29: 'Ranged Haste',
	30: 'Spell Haste',
	31: 'Hit',
	32: 'Crit',
	35: 'Resilience',
	36: 'Haste',
	37: 'Expertise',
	38: 'Attack Power',
	39: 'Ranged Attack Power',
	42: 'Spell Power',
	43: 'Mana Regen',
	44: 'Armor Penetration',
	45: 'Spell Power',
	47: 'Spell Penetration',
	48: 'Block Value',
};

export function reforgeStatTypeName(statType: number): string {
	return reforgeStatTypeNames[statType] || `Stat ${statType}`;
}

// The stat types a config can sensibly list: the ones the sim has a stat for, which is the same
// set reforgeStatTypeToStats covers in sim/core/reforging.go.
export const KNOWN_REFORGE_STAT_TYPES = Object.keys(reforgeStatTypeNames).map(Number);

// serverStatValue is LoadItemStatInfo plus FindItemStat: the first item_template row of that type
// with a value above zero, else 0.
function serverStatValue(item: Item, statType: number): number {
	return item.serverStats.find(stat => stat.statType == statType && stat.value > 0)?.value ?? 0;
}

// reforgeAmount is CalculateReforgePct: the configured percentage of the template stat, floored, in
// float32 like the server.
export function reforgeAmount(item: Item, fromStatType: number, config: Reforging = LIVE_REFORGING): number {
	const value = serverStatValue(item, fromStatType);
	if (value <= 0) {
		return 0;
	}
	return Math.floor(Math.fround(Math.fround(value) * Math.fround(Math.fround(config.percentage) / 100)));
}

// isValidReforge follows ItemReforge::IsReforgeable and ::Reforge, which read the item_template
// stats rather than what the sim made of them. A stat type the sim has no stat for is out: core's
// ReforgeStats drops the pair, so offering it here would show a reforge the sim then ignores.
export function isValidReforge(item: Item, reforge: ItemReforge | null | undefined, config: Reforging = LIVE_REFORGING): boolean {
	return (
		!!reforge &&
		config.enabled &&
		config.statTypes.length > 0 &&
		item.serverStats.length > 0 &&
		item.serverStats.length < MAX_ITEM_PROTO_STATS &&
		reforge.fromStatType != reforge.toStatType &&
		config.statTypes.includes(reforge.fromStatType) &&
		config.statTypes.includes(reforge.toStatType) &&
		KNOWN_REFORGE_STAT_TYPES.includes(reforge.fromStatType) &&
		KNOWN_REFORGE_STAT_TYPES.includes(reforge.toStatType) &&
		serverStatValue(item, reforge.toStatType) == 0 &&
		reforgeAmount(item, reforge.fromStatType, config) >= 1
	);
}

export function validReforges(item: Item, config: Reforging = LIVE_REFORGING): Array<ItemReforge> {
	return config.statTypes
		.flatMap(fromStatType => config.statTypes.map(toStatType => ItemReforge.create({ fromStatType, toStatType })))
		.filter(reforge => isValidReforge(item, reforge, config));
}

// e.g. "33 Crit → Haste"
export function reforgeLabel(item: Item, reforge: ItemReforge, config: Reforging = LIVE_REFORGING): string {
	return `${reforgeAmount(item, reforge.fromStatType, config)} ${reforgeStatTypeName(reforge.fromStatType)} → ${reforgeStatTypeName(
		reforge.toStatType,
	)}`;
}
